File: kaiix/tree/BinaryTreeLevelOrderTraversal.java
package tree;

import java.util.ArrayList;

/*
 * http://oj.leetcode.com/problems/binary-tree-level-order-traversal/
 */
public class BinaryTreeLevelOrderTraversal {
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) { val = x; }
    }

    public static ArrayList<ArrayList<Integer>> levelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root != null) {
            ArrayList<TreeNode> siblings = new ArrayList<TreeNode>();
            siblings.add(root);
            while (true) {
                final int count = siblings.size();
                if (count == 0) {
                    break;
                }
                ArrayList<Integer> siblingsValue = new ArrayList<Integer>();
                for (int i = 0; i < count; i++) {
                    TreeNode node = siblings.get(0);
                    siblingsValue.add(node.val);
                    if (node.left != null) siblings.add(node.left);
                    if (node.right != null) siblings.add(node.right);
                    siblings.remove(0);
                }
                result.add(siblingsValue);
            }

        }
        return result;
    }

    public static void main(String []args){
        TreeNode root = new TreeNode(3);
        root.left = new TreeNode(9);
        root.right = new TreeNode(20);
        root.right.left = new TreeNode(15);
        root.right.right = new TreeNode(7);

        System.out.println(levelOrder(root));
    }
}

File: dave/tree/BinaryTreeLevelOrderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrder(self, root):
        q = [[root]]
        ret = []
        while q:
            l = q.pop(0)
            v = []
            ll = []
            for n in l:
                if n:
                    v.append(n.val)
                    ll.append(n.left)
                    ll.append(n.right)
            if v:
                ret.append(v)
            if ll:
                q.append(ll)
        return ret
